rust/lance/src/dataset/cleanup.rs

```diff
@@ -201,6 +204,42 @@ impl<'a> CleanupTask<'a> {
         Ok(())
     }
 
+    async fn check_unreferenced_files_are_not_tagged(
+        &self,
+        inspection: &CleanupInspection,
+    ) -> Result<()> {
+        let tags = self.dataset.tags().await?;
+
+        let unreferenced_versions: HashSet<u64> = HashSet::from_iter(
+            inspection
+                .old_manifests
+                .iter()
+                .map(|p| parse_version_from_path(p).unwrap()),
+        );
+
+        let unreferenced_tags: HashMap<String, u64> = tags
+            .iter()
+            .filter_map(|(k, &v)| {
+                if unreferenced_versions.contains(&v.version) {
+                    Some((k.clone(), v.version))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        match unreferenced_tags.len() {
+            0 => Ok(()),
+            _ => Err(Error::Cleanup {
+                message: format!(
+                    "{} tagged version(s) have been marked for cleanup. Please delete the following tag(s) to enable cleanup: {:?}",
```

Comment:
I wonder if we should have a setting for the behavior on this. It would be nice to have a setting to just warn instead of error if there are tags, and just filter out the versions that are tagged from cleanup candidates. That is, if there are versions 1, 2, 3, 4, and we are deleting before 4, but 2 is tagged, it should only delete 1 and 3. What do you think of that?

Reply:
I wasn't sure about this design decision and could be persuaded either way. I thought about adding a flag to control this behaviour, but I wanted to keep the API simple. I've gone with the more conservative option for now, but it would be really useful to keep a tagged version while deleting up the others.

Would you like a flag to allow deleting untagged versions when a tagged version is present, or should we just make this the default behaviour and never return a Error::CleanupError?

Reply:
Maybe `ignore_tagged` is a better keyword than `delete_untagged`? Because either way we might delete the untagged ones. It's really about whether or not we are erroring if there are tags.

One of the motivations I have in mind is in the near future we will have some sort of auto-cleanup run after each write operation. This would automatically clean up versions older than some given time. It would be really annoying if it failed when there are tags.

Reply:
Yep, that's a better name. Sounds good!

Reply:
@wjones127 I started implementing this today. Upon implementing I realised this would be a breaking change as it requires adding an argument to `cleanup_old_versions`. So far, all of the changes have been backwards compatible. So, do we still want to add a flag for returning an error, or should we simply log a warning to the user? We would still delete untagged old versions as you proposed.

If we do decide to add a flag, I wonder whether we should call it something more explicit like `error_if_old_tags` as I found myself getting a bit confused by `ignore_tagged` during implementation (as it sounds like, if set to `false` the tagged versions should be cleaned up). What do you think?

Reply:
Adding an argument is fine. That's a kind of breaking change we generally don't have any issue with.

Reply:
I decided to add the breaking change with the argument `error_if_old_versions_tagged`. It may get annoying to constantly suppress warnings if the user has no intention of ever removing the old tags. I've update the APIs and the tests to reflect the new argument.